sdks/go/examples/complete/tfidf/tfidf.go

```diff
@@ -0,0 +1,272 @@
+	s = s.Scope("ComputeTfIdf")
+
+	// Get all filenames
+	values := beam.DropValue(s, uriContent)
+	// Distinct the filenames
+	distinctUris := filter.Distinct(s, values)
+	// Add fixed key, drop value, count, then drop key to get the number
+	// of distinct filenames. This equivalent is Count.globally() in Java.
+	// documentCount is then a singleton PCollection<int>
+	distinctUrisWithKey := beam.AddFixedKey(s, distinctUris)
+	allFixedKey := beam.DropValue(s, distinctUrisWithKey)
+	fixedKeyCount := stats.Count(s, allFixedKey)
+	documentCount := beam.DropKey(s, fixedKeyCount)
+
+	// Get each word and its corresponding file.
+	// Use the WordFile struct to storee this so that we can avoid KV's
+	uriToWords := beam.ParDo(s.Scope("SplitWords"), func(uri, line string, emit func(WordFile)) {
+		regex := regexp.MustCompile(`\W+`)
+		words := regex.Split(line, -1)
+		for _, word := range words {
+			if strings.EqualFold(love, word) {
+				log.Infof(context.Background(), "Found %s", strings.ToLower(word))
+			}
+			if word != "" {
+				emit(WordFile{
+					File: uri,
+					Word: strings.ToLower(word),
+				})
+			}
+		}
+	}, uriContent)
+
+	// Get distinct (word, file) pairs
+	distinctUriWords := filter.Distinct(s, uriToWords)
+	// Get just the words
+	words := beam.ParDo(s.Scope("GetWords"), func(wordFile WordFile) string {
+		return wordFile.Word
+	}, distinctUriWords)
+	// Compute a mapping from each word to the total
+	// number of documents in which it appears.
+	wordToDocCount := stats.Count(s, words)
+
+	// Get just the filenames
+	uris := beam.ParDo(s.Scope("GetFilenames"), func(wordFile WordFile) string {
+		return wordFile.File
+	}, uriToWords)
+	// Compute a mapping from each URI to the total
+	// number of words in the document associated with that URI.
+	uriToWordsTotal := stats.Count(s, uris)
+
+	// Count, for each (URI, word) pair, the number of
+	// occurrences of that word in the document associated
+	// with the URI.
+	uriAndWordToCount := stats.Count(s, uriToWords)
+
+	// Adjust the above collection to a mapping from
+	// (URI, word) pairs to counts into an isomorphic mapping
+	// from URI to (word, count) pairs, to prepare for a join
+	// by the URI key.
+	uriToWordAndCount := beam.ParDo(s.Scope("SwitchKeys"), func(wordFile WordFile, instances int) (string, WordCount) {
+		return wordFile.File, WordCount{
+			Word:      wordFile.Word,
+			Instances: instances,
+		}
+	}, uriAndWordToCount)
+
+	// Compute a mapping from each word to a (URI, term frequency)
+	// pair for each URI. A word's term frequency for a document
+	// is simply the number of times that word occurs in the document
+	// divided by the total number of words in the document.
+	uriToWordAndCountAndTotal := beam.CoGroupByKey(s, uriToWordsTotal, uriToWordAndCount)
+	wordToUriAndTermFrequency := beam.ParDo(s.Scope("ComputeTermFrequencies"), func(filename string, totalDocCountValues func(*int) bool, wordCountValues func(*WordCount) bool, emit func(string, FileFrequency)) {
```

Comment:
Note: while Anonymous functions do work in some cases, they're impossible to optimize WRT reflection. (Which adds per element overhead) they also require looking up the function pointer in the Debug symbol table, which isn't necessarily guaranteed to work in later go versions.

Consider making this a named function so it can be tested too.